### DISCUSSION TITLE
[4.0] Web services - Can use some help with exception handling

### DISCUSSION THREAD

**Discussion by pjdevries** [joomla/joomla-cms#34432] [Q&A]
I am a bit confused about web service exception handling:

- Why are there a fixed, limited number of handlers that need to be registered explicitly?
- Why is it that some of the handlers do nothing more than simply echo the HTTP response description, which is already clear from the status code, and completely ignore the Exception message?
- `JsonapiRenderer::render()` registers a limited set of handlers. What would be the proper way to add custom handlers?

@wilsonge appears to be the originator of the general implementation, so I hope he can shed some light on the concept, allowing me to push forward with my media web service implementation (#34314).

**Comment by wilsonge** [comment]
> Why are there a fixed, limited number of handlers that need to be registered explicitly?

The intention is to genericise them and allow extensions to add their own custom handlers. The registering of handlers is just how the underlying library I chose to use to make life easier works (although given that entire library is now deprecated 4 years later - unsure how much value it really gives us).

The original handlers were mainly extensions of ones' that were available in the parent library. Then when we did GSOC 2019 my student had nowhere else to add them for more component specific things - so things kinda got splashed around. I needed to clean up some like the SendMailException into com_contact specifically.

> JsonapiRenderer::render() registers a limited set of handlers. What would be the proper way to add custom handlers?

I was thinking about this a few weeks ago. I suspect registering the handler class in the DI container and then allowing the extensions child containers to register custom handlers within.

> Why is it that some of the handlers do nothing more than simply echo the HTTP response description, which is already clear from the status code, and completely ignore the Exception message?

Probably just my shoddy coding to be honest. It's been a few years now but I don't remember any specific reason. Probably just with some of the exception handlers (e.g. `NotAcceptable`) I didn't forsee anyway that people would actually inject any useful text within.

**Reply by pjdevries** [reply]
@wilsonge Thanx for the feed back.

> > JsonapiRenderer::render() registers a limited set of handlers. What would be the proper way to add custom handlers?
> 
> I was thinking about this a few weeks ago. I suspect registering the handler class in the DI container and then allowing the extensions child containers to register custom handlers within.

Not something I feel confident enough to take on at this moment. I would first need to study usage of the DI container a bit.

> > Why is it that some of the handlers do nothing more than simply echo the HTTP response description, which is already clear from the status code, and completely ignore the Exception message?
> 
> Probably just my shoddy coding to be honest. It's been a few years now but I don't remember any specific reason. Probably just with some of the exception handlers (e.g. `NotAcceptable`) I didn't forsee anyway that people would actually inject any useful text within.

Maybe I can find the time and courage to improve on your code later on :)
